mwdb/resources/oauth.py

```
@@ -0,0 +1,238 @@
+        """
+        ---
+        summary: Login using OpenID Connect
+        description: |
+            TODO
+        tags:
+            - auth
+        parameters:
+            - in: path
+              name: provider_name
+              schema:
+                type: string
+              description: OpenID provider name.
+        responses:
+            200:
+                description: Returns authorization URL for chosen provider
+                content:
+                  application/json:
+                    schema: OpenIDLoginResponseSchema
+        """
+        provider = (
+            db.session.query(OpenIDProvider)
+            .filter(OpenIDProvider.name == provider_name)
+            .first()
+        )
+        if not provider:
+            raise NotFound(f"Requested provider name '{provider_name}' not found")
+
+        redirect_uri = f"{app_config.mwdb.base_url}/oauth/authorize"
+        url, state, nonce = provider.create_authorization_url(redirect_uri)
+
+        schema = OpenIDLoginResponseSchema()
+        return schema.dump({"authorization_url": url, "state": state, "nonce": nonce})
+
+
+class OpenIDAuthorizeResource(Resource):
+    def post(self, provider_name):
+        provider = (
+            db.session.query(OpenIDProvider)
+            .filter(OpenIDProvider.name == provider_name)
+            .first()
+        )
+        if not provider:
+            raise NotFound(f"Requested provider name '{provider_name}' not found")
+
+        schema = OpenIDAuthorizeRequestSchema()
+        obj = loads_schema(request.get_data(as_text=True), schema)
+        redirect_uri = f"{app_config.mwdb.base_url}/oauth/authorize"
+        userinfo = provider.fetch_id_token(
+            obj["code"], obj["state"], obj["nonce"], redirect_uri
+        )
+        # 'sub' bind should be used instead of 'name'
+        identity = OpenIDUserIdentity.query.filter(
+            OpenIDUserIdentity.sub_id == userinfo["sub"]
+        ).one()
```

There are few problems here:
- We don't check `provider_id`, so there's possible `sub` collision across different providers
- `OpenIDUserIdentity.query` vs `db.session.query(OpenIDUserIdentity)` inconsistency
- `.one()` throws `NoResultFound` instead of returning None - use `first()`
```suggestion
        identity = db.session.query(OpenIDUserIdentity).filter(
            OpenIDUserIdentity.sub_id == userinfo["sub"],
            OpenIDUserIdentity.provider_id == provider.id,
        ).first()
```